src/python/createrepo_c/__init__.py

```diff
@@ -358,6 +362,273 @@ def decompress_file(src, dst, comtype, stat=None):
+        self.working_metadata_files["filelists"].writer.add_pkg(pkg)
+        self.working_metadata_files["other"].writer.add_pkg(pkg)
+
+        if self._with_sqlite:
+            self.working_metadata_files["primary_db"].writer.add_pkg(pkg)
+            self.working_metadata_files["filelists_db"].writer.add_pkg(pkg)
+            self.working_metadata_files["other_db"].writer.add_pkg(pkg)
+
+    def add_repomd_metadata(self, name, path):
+        assert not self._finished, self._FINISHED_ERR_MSG
+
+        shutil.copy2(path, self.repodata_path)
+        self.additional_metadata_files[name] = path
+
+    def add_update_record(self, rec):
+        assert not self._finished, self._FINISHED_ERR_MSG
+
+        # lazily create the updateinfo entry
+        if "updateinfo" not in self.working_metadata_files:
+            upd_xml_path = Path(self.repodata_path) / (self.upd_xml_name + self._compression_suffix)
+            self.working_metadata_files["updateinfo"] = MetadataInfoHolder(
+                upd_xml_path, UpdateInfoXmlFile(str(upd_xml_path), compressiontype=self._general_compression)
+            )
+
+        self.working_metadata_files["updateinfo"].writer.append(rec)
+
+    def add_content_tag(self, tag):
+        assert not self._finished, self._FINISHED_ERR_MSG
+        self._repomd.add_content_tag(tag)
+
+    def add_distro_tag(self, tag, cpeid=None):
+        assert not self._finished, self._FINISHED_ERR_MSG
+        self._repomd.add_distro_tag(tag, cpeid=cpeid)
+
+    def add_repo_tag(self, tag):
+        assert not self._finished, self._FINISHED_ERR_MSG
+        self._repomd.add_repo_tag(tag)
+
+    def finish(self):
+        assert not self._finished, self._FINISHED_ERR_MSG
+        self._finished = True
+
+        metadata_with_sqlite = {
+            "primary": "primary_db",
+            "filelists": "filelists_db",
+            "other": "other_db",
+        }
+
+        records = []
+
+        # Create all the repomdrecords for the standard metadata
+        for record_name, metadata_info in self.working_metadata_files.items():
+            sqlite_name = metadata_with_sqlite.get(record_name)
+
+            # Close all of the metadata files being actively edited
+            if record_name not in metadata_with_sqlite.values():
+                metadata_info.writer.close()
+                record = RepomdRecord(record_name, str(metadata_info.path))
+                record.fill(self._metadata_checksum_type)
+                records.append(record)
+
+                if sqlite_name and self._with_sqlite:
+                    sqlite_metadata_info = self.working_metadata_files[sqlite_name]
+                    sqlite_metadata_info.writer.dbinfo_update(record.checksum)
+                    sqlite_metadata_info.writer.close()
+            else:
+                record = RepomdRecord(record_name, str(metadata_info.path))
+                record = record.compress_and_fill(self._metadata_checksum_type, BZ2_COMPRESSION)
```

Comment:
While trying out the `RepositoryWriter` I noticed the created repositories are invalid. I think this section specifically exposes a bug in the library. 

`record.compress_and_fill(self._metadata_checksum_type, BZ2_COMPRESSION)` returns a new record with filled in checksums while the old one (created on the line above - 603) is freed. However these two lines: https://github.com/rpm-software-management/createrepo_c/blob/master/src/repomd.c#L557-L558 make the new record use chunk from the old (freed) one. This leads to garbage values.
Could this be fixed as a part of this PR?

Reply:
@kontura It might take a while before this PR is ready depending on how much time I can dedicate to it.  It should probably go in a separate PR.  I can maybe take a shot at fixing it next week though.

Reply:
Ok, that is no problem.

I think we just need to change `record->chunk` to `crecord->chunk` for both of those lines.  :slightly_smiling_face: